Added Objective Obstacle Functionality

MarvelProgramming opened the pull request:
-Added dev_fundamentals code snippet which changes the layer ordering of specific objects, such that the player renders beneath them
-Updated initial state to account for the destroyedEntities, unlockedTransitions, hiddenEntities, etc, of individual maps under "lovelace_tower"

For issue #46

Commit:
Added Objective Obstacle Functionality

-Added dev_fundamentals code snippet which changes the layer ordering of specific objects, such that the player renders beneath them
-Updated initial state to account for the destroyedEntities, unlockedTransitions, hiddenEntities, etc, of individual maps under "lovelace_tower"

MarvelProgramming commented:
~~I noticed that @nmadadair was assigned to the issue for this PR after completing it. If she has any outstanding changes, I'm happy with accepting those in place of this!~~

\*UPDATE\*: @nmadadair gave me the all clear!

rmkubik commented:
We should ensure this does not conflict with the scripting changes @nmadadair and I were doing today (11/29). The logic in this PR seems pretty distinct. However, I am worried about git getting confused by the old code deletions.

rmkubik commented on levels/lovelace_tower/events.js:
```diff
@@ -1,353 +1,162 @@
+
+  // Top object pieces need to be longer so they always depth sort
+  // above the player, even when the player would normally be depth
+  // sorted above them.
```
They need to be above the player even when they'd be above the player?

rmkubik replied:
I think we might be able to use the newer [zIndex](https://github.com/twilio/twilioquest/blob/3.2_next/src/js/app/components/game/DynamicLevel.js#L294-L298) property for this usecase now.

I think this is used in TOIK exterior level on layers. I believe it can be applied to objects directly as well though.

MarvelProgramming replied:
> They need to be above the player even when they'd be above the player?

This is to say, render top objects above the player even when they normally wouldn't. Bit of a lazily edited version of [this](https://github.com/TwilioQuest/twilioquest-dev-fundamentals/blob/main/levels/tower_of_knowledge/events.js#:~:text=//%20Top%20shelf%20pieces,sorted%20above%20them.) haha


> I think we might be able to use the newer [zIndex](https://github.com/twilio/twilioquest/blob/3.2_next/src/js/app/components/game/DynamicLevel.js#L294-L298) property for this usecase now.
> 
> I think this is used in TOIK exterior level on layers. I believe it can be applied to objects directly as well though.

Just tried it and it works! I'll go ahead and use that approach instead

rmkubik commented on levels/lovelace_tower/events.js:
```diff
@@ -1,353 +1,162 @@
+  ) {
+    unlockPairs.forEach(([objectiveKey, hiddenDoorKey]) => {
+      if (event.objective === objectiveKey) {
+        worldState.insideLovelaceTower.hiddenEntities.push(hiddenDoorKey);
```
This will push duplicates of the hiddenDoorKey each time the objective is completed right? We might want to check if it's already there first:
```js
if (!worldState.insideLovelaceTower.hiddenEntities.includes(hiddenDoorKey)) {

}
```

MarvelProgramming replied:
You're absolutely right! I pulled that from [dev fundamentals](https://github.com/TwilioQuest/twilioquest-dev-fundamentals/blob/main/levels/tower_of_knowledge/events.js#:~:text=%5D%3B-,if%20\(,%7D,-//%20Hide%20entities), so we'll want to update that too when we have the time. I'll make a GitHub Issue for it as a reminder.

\*UPDATE\*: All done!

rmkubik commented on levels/lovelace_tower/events.js:
```diff
@@ -1,353 +1,162 @@
-      world.enablePlayerMovement();
+    doorPairs.forEach(([objectiveKey, doorKey]) => {
+      if (event.objective === objectiveKey) {
+        worldState.insideLovelaceTower.openedDoors.push(doorKey);
```
Same as prev comment!

rmkubik commented on levels/lovelace_tower/events.js:
```diff
@@ -1,353 +1,162 @@
-  }
+  // If Yes:
+  // [] Sparkle animation appears on exit door and is interactable
+  // [] Previous entry door back into Tower becomes locked
```
Is this all the pre-existing comment stuff?

MarvelProgramming replied:
Yep! I believe @nmadadair has been using them to keep track of things, if I'm not mistaken

nmadadair replied:
Comments are mine, but I'm deleting them as the code gets written. Regarding old inside_perimeter event.js code, we've deleted it all and are starting from scratch for both new levels.

rmkubik reviewed (commented):
Couple small questions/comments here

MarvelProgramming commented:
> We should ensure this does not conflict with the scripting changes @nmadadair and I were doing today (11/29). The logic in this PR seems pretty distinct. However, I am worried about git getting confused by the old code deletions.

Are we looking to reuse some of what's in the old code (aside from how we would via [this](https://github.com/TwilioQuest/twilioquest-api-academy/pull/78))? Pretty much all of it was copied over from inside_perimeter's [event.js file](https://github.com/TwilioQuest/twilioquest-api-academy/blob/main/levels/inside_perimeter/events.js). I can always rewrite this stuff around the changes you two made if need be!

Commit:
Fixed Duplicate Key Bug

-Added check for whether hiddenDoorKey exists in "worldState.insideLovelaceTower.hiddenEntities" before adding it
-Added check for whether doorKey exists in "worldState.insideLovelaceTower.openedDoors" before adding it

rmkubik commented:
I think you're correct! I forgot that we were working on house_ceremony events.js yesterday.

If you'll handle merging, @MarvJohnson, i think this into #81 afterward this PR should be good to go!

I don't want to have Nicole worry about merging this into her work if you can handle it. In the new year, we'll want to do some team git training so that we can all handle it. :-)

MarvelProgramming committed:
Merge branch 'milestone-1-aa3' into 46-implement-lovelace-tower-progress-barriers

MarvelProgramming commented on levels/lovelace_tower/events.js:
```diff
@@ -11,16 +66,73 @@ const INITIAL_STATE = {
   console.log(worldState);
 
-// Library exit not accessible until obj5Complete === true
-world.disableTransitionAreas("exit_to_library_corridor");
```
Removed this because of how we're handling the opening of these doors. Based on [this](https://github.com/TwilioQuest/twilioquest-api-academy/issues/71#:~:text=Lovelace%20Tower%3A%20When%20the%20player%20performs%20the%20spell%20on%20the%20blue%20door%20at%20the%20end%20of%20Lovelace%20Tower%2C%20after%20the%20spell%20animation%20plays%2C%20the%20door%20should%20open%20and%20trigger%20the%20levelChange%20to%20the%20secret%20library.) and [this](https://github.com/TwilioQuest/twilioquest-api-academy/issues/71#:~:text=Secret%20Lovelace%20Library%3A%20When%20the%20player%20performs%20the%20spell%20on%20the%20blue%20exit%20door%20inside%20Lovelace%20Library%2C%20after%20the%20spell%20animation%20plays%2C%20the%20door%20should%20open%20and%20trigger%20the%20levelChange%20back%20to%20the%20Lovelace%20corridor.), we're looking to have the player cast a spell on the door before it opens and they're able to move on to the next level. This is being handled through a combination of [property binding](https://twilioquest.github.io/extension-docs/api/custom_objects.html#property-binding) and the restructured spell code (examples shown below).

<img width="1440" alt="property-binding-example" src="https://user-images.githubusercontent.com/51891409/204936303-c5dc96c9-a77f-4153-b24a-dcebf5576c8a.png">

<img width="1057" alt="restructured-spell-example" src="https://user-images.githubusercontent.com/51891409/204936541-4c259b0b-28c4-4386-bff2-cd25c6c618c2.png">

MarvelProgramming commented on levels/lovelace_tower/events.js:
```diff
@@ -46,17 +158,6 @@ if (event.name === "triggerAreaWasEntered" && event.target.key === "objectiveBar
 };
 
-// Remove objective barriers
-if (worldState.obj1Complete === true){
```
Removed these because this logic is being handled in some of the code from this PR. That being said, I imagine we could use [property binding](https://twilioquest.github.io/extension-docs/api/custom_objects.html#property-binding) for this sort of thing too? Unless we wanted fancy fading/sliding animations, it'd probably be easier in the long run and reduce the amount of repeat code

Commit:
Fixed Obstacle Render Order Bug

-Added zIndex to the "top" parts of each obstacle with a value of 1, such that they render above the player

Commit:
Fixed Bugs

-Added "disabled" property to "exit_to_lovelace_library" levelChange and set it to true by default
-Updated "exit_to_lovelace_library" levelChange position to be higher, such that the player would have to walk "into" the doorway before changing levels
-Fixed reference exception that would throw because the worldState was being accessed incorrectly
